### ISSUE TITLE
Alter extraction code to only extract "new" scenarios, updated data

### ISSUE THREAD

**Issue by denised** [ProjectDrawdown/solutions#381]
As the Excel workbooks evolve, the extraction code needs to evolve with them.  This poses a challenge for scenarios that were created with the older versions of the workbooks and may no longer behave as they originally did as a result.  One part of the solution is to capture the actual data that was saved with the scenario, which is issue #258 

For workbooks that have already been extracted before, we also have a unique opportunity to more accurately preserve the historical intent of these scenarios: we have the existing extractions of them, that were created from older workbooks.  We should not overwrite these previously-extracted scenarios, when re-extracting a workbook to get new scenarios.

This might just be implemented by documented procedure (extract into a new directory, copy over new scenarios only), or we might automate some of it.

Some investigation should be done to figure out a good solution for this.   Looking closely at how __init__.py is influenced by scenarios will be important, since __init__.py must function regardless of the scenario pedigree.

**Comment by denised** [comment]
This would also apply to expected results.  We need the ability to make sure that we capture the expected results only for the new scenarios that we are extracting, and merge them with the existing expected results.

**Comment by johnpaulalex** [comment]
I can take a shot at this.

**Comment by denised** [comment]
@johnpaulalex When you get a good grasp on how this is going to go, we should take a moment to talk about how to update VMAs.  

For one thing, I want to move the VMA dictionary into a JSON file, similar to the adoptions and tam files, so it would be helpful if I did that before you got to needing to update it :-)

For another, we need to update some parameters or metadata attached to VMAs, because they may have changed and because we weren't extracting some things, like the bound_correct parameter I just added in [ddda5e9](https://github.com/ProjectDrawdown/solutions/commit/ddda5e9a34d4732d4a263a75f99ee1149a7340e3)

**Comment by johnpaulalex** [comment]
I did a quick overview of what's exported from solution_xls_extract[1].  It sounds like the most primitive use case is that we just run it as-is and it overwrites everything and we rely on the caller to filter out what they want before they commit - but iiuc you'd like to add some optional smarts? 

iiuc there could be problems where the export will fail if some scenarios are no longer valid (how?) so we'd want to skip them? I could make a simple flag to skipping specific scenarios; or a smarter flag that skips the already-exported scenarios...would this only apply to the ac/ files? 

I'm not sure what's needed for the shared data (iiuc everything outside of ac?) - can we just overwrite it and assume the latest scenarios will work with it? Do we need to somehow stow the old shared data (which can never be re-exported iiuc) so older scenarios can still work? This makes me wonder if all the shared data should in fact be duplicated per scenario; or each scenario should reference which version of the shared data it should use...

[1] there's the exporting of __init__.py with a lot of dynamic content in the generated code: dynamic vmas, units, tam/tla, adoption, helpertables, emissionfactors, unit adoption, first cost, operating cost, co2 and ch4 calcs.

Then there's the other data files exported:
ac/*.json (one per scenario)
vmas
custom tlas
adoption (including custom)

**Comment by denised** [comment]
First off, thank you for a nice commit to get #400 closed! :-)

My suggestion: make a copy of solution_xls_extract.py to something appropriate like solution_xls_update.py, and then feel free to completely hack at that code to retain only the parts you need to do updating.   Scenario updating shouldn't touch the __init__.py code, it should only update the data files, as you describe.

I am not worried about keeping history, for two reasons:  (1) the PD researchers aren't worried about it (they delete old scenarios when they stop working) and (2) We have the history in git.  If at some point someone needs to resurrect a scenario and its associated data files as it used to be, that is how we'll handle it.

So yes, overwrite all data files.  You could  also delete data files that aren't present in the Excel any longer (there might be old custom adoptions, etc.)   As we've previously discussed, we do not want scenarios that we've already deleted (whose creation_date was less than the maximum creation_date of the scenarios we do have).

The other complexity is the new file we have that didn't exist before: the .json "directory" files that used to be code in the __init__ listing the adoption or tam sources.   These need to be updated as files now.

A couple of thoughts to help scope this:
 * The existing solution_xls_extractor.py hasn't been updated to all the code changes, and you don't need to fix this at this time.  When we decide we need a fixed version of it (which won't be any time soon), we'll take your new code and worry about merging them.[1]
 * I don't expect that updating will go seamlessly.  _Most_ of the time, logic in the spreadsheet hasn't changed between one version and the next, but sometimes it has.  In that case, test results will fail until we figure out what has changed and make the corresponding changes to __init__.py  That one-by-one investigation isn't part of the scope of this work item (once you've gotten past the investigation that the update was performed correctly).   We'll handle those as additional "fix broken models" issues.    You'll probably fix a few along the way as you figure out what's going on, but just know you don't have to.
 * We'll have to figure out the issue for creating updated expected.zip files.   One option is for me to just go do these manually, or the other option is to resurrect Denton's code for this --- I believe that code works on a Mac, but it did not work for me on a Windows machine.  (The advantage to Denton's code was that it was automatable, which my version isn't.)   I'll certainly create a few  for you to work with for now.

And I can't say this enough:  thank you for your help!   I'm very aware that this is a lot of work, and I really appreciate the time you are investing in this project!

[1] I realize this sounds horrible from a software engineering perspective.  It is only coming from the priority to get the existing models updated sooner rather than later.

-- 
  Denise Draper
  ***@***.***

**Comment by johnpaulalex** [comment]
Happy to help!

How about these mods to the existing extract script - doesn't seem like forking to a new script would help here although I will keep that in mind:
   * After it does the mkdir for the specific solution, read that dir in case it's already present and populated:
       *  Read existing ac/*.json scenario names and creation_dates, if any; stow earliest creation_date.
   * Then do some targeting cleaning out of solution subdirs - e.g. various <subdir>/*.csv and *.json patterns.  This removes the need for more fancy removal of unused data files - just get rid of them all. The globs should help avoid deleting any random extra files that should stick around(or would deleting */* be more helpful?). Either way, if the solution is brand-new, this is a no-op.
   * Finally, write out all files as before, filtering scenarios with an earlier creation_date.
I could skip overwriting the __init__.py but I would think it's helpful to have the brand-new code as a reference -- you can always just do a full restore of the file if necessary, but if some parts need updating you can pull them in. 

iiuc this would make updating pretty similar to starting from scratch: you run the same script as always, then you delete any new ac/*.json that you don't want, you merge __init__.py (this is new) and update PDS1/2/3 if needed, and then you're done (well, then you turn to fighting the tests, expected.zip, etc...)

**Comment by denised** [comment]
I agree that it is a good idea not to delete random files.  They may be part of customized code specific to that solution.

If you want to generate updated __init__.py, I guess I I can't complain :-)   I think it would be good if new file were named something like new__init__.py so that the workflow would be that someone has to diff them before committing the changes.

In addition to the changes in the code format, there are new quirks parameters. The existing code was clever about guessing what the settings of these should be, based on analyzing the current crop of models (e.g. Denton would observe that the current set of Electricity models all had some feature in common, and put a test for that in the code).  That method inherently doesn't age well.  So my instinct is do less of that, and leave it to manual examination to set properly.  (And if the spreadsheets haven't been changed on update, then the old quirks settings should still be good anyway.)
So this is a judgment call on your part if you want to try to set the new quirks parameters in the generated code.

-- 
  Denise Draper
  ***@***.***
